### PR TITLE
test for KanesMethod to detect nonlinearity

### PR DESCRIPTION
**Pull request by Peter230655**
#### Release Notes

* physics.mechanics
  * This is a test function to check whether my proposed changes to KanesMethod will detect a non linearity in the given velocity constraints.

### CODE REVIEW

**Comment by Peter230655** [comment]
It failed the symengine test, these are the details.
I am not clear, what this means: the current KanesMethod will not raise a ValueError.

image

**Comment by Peter230655** [comment]
It also failed the test pull request.
I looked at the details, but no idea what they mean.
Any help is greatly appreciated!

image

**Comment by oscarbenjamin** [comment]
The error message says "did not raise" because in the test code you have used `with pytest.raises` which asserts that some exception should be raised. Apparently no exception was raised when the test was run.

**Comment by Peter230655** [comment]
> The error message says "did not raise" because in the test code you have used `with pytest.raises` which asserts that some exception should be raised. Apparently no exception was raised when the test was run.

But this is the Catch-22: I proposed a change in KanesMethod to raise a value error if nonlinear speed constraints are presented.
The current KanesMethod does not raise a ValueError when nonlinear speed constraints are presented - but likely the OEMs will be wrong. Surely this test is with the current KanesMethod, so no value error is raised.
What to do?

Thanks!!

**Comment by oscarbenjamin** [comment]
It sounds like you have written the test code correctly but now to make the test pass you need to go and fix the Kane's method code that isn't checking linearity properly. Once that is done the test will pass.

Not everyone subscribes to this but there is a programming model called test-driven development (TDD). In TDD it is suggested that the way to write tested code is:

1. Write a test for the code first before writing the actual code.
2. Run the test and confirm that it fails.
3. Now write the code (fix the bug etc) that is supposed to make the test pass.
4. Run the test and confirm that it passes.

The idea is that you confirm that the test fails before trying to fix the problem so that you can see that the test code itself is working (it's purpose is to fail until the problem is fixed!).

I might be wrong but it sounds like you have completed steps 1 and 2 and now you need to work on 3 and 4.

I haven't actually checked that the test code is correct though. Jason would have to do that because I don't know the Kane's method code.

**Comment by Peter230655** [comment]
> It sounds like you have written the test code correctly but now to make the test pass you need to go and fix the Kane's method code that isn't checking linearity properly. Once that is done the test will pass.
> 
> Not everyone subscribes to this but there is a programming model called test-driven development (TDD). In TDD it is suggested that the way to write tested code is:
> 
> 1. Write a test for the code first before writing the actual code.
> 2. Run the test and confirm that it fails.
> 3. Now write the code (fix the bug etc) that is supposed to make the test pass.
> 4. Run the test and confirm that it passes.
> 
> The idea is that you confirm that the test fails before trying to fix the problem so that you can see that the test code itself is working (it's purpose is to fail until the problem is fixed!).
> 
> I might be wrong but it sounds like you have completed steps 1 and 2 and now you need to work on 3 and 4.
> 
> I haven't actually checked that the test code is correct though. Jason would have to do that because I don't know the Kane's method code.

Thanks!

1. Indeed I wrote the code to change KanesMethod long before I knew I had to write a test to check my proposed change for KanesMethod.
2. Writing the test was also trivial, except the actual checking whether a ValueError was raised. Here Jason showed me how to do it.
3. Actually both, my proposed change to KanesMethod, and the test whether it works are trivial - and I tested them locally many, many times. These Github issues, **where you and Jason have helped me so much!** are FAR more difficult for me

In summary: I am convinced both work - but how do I get the test to pass the Github hurdles?

**Comment by oscarbenjamin** [comment]
The problem is that now in this new PR there are no changes to the Kane's method code. Look in the "files changed" tab above. Here only the test has been added but the bug has not been fixed.

**Comment by Peter230655** [comment]
> The problem is that now in this new PR there are no changes to the Kane's method code. Look in the "files changed" tab above. Here only the test has been added but the bug has not been fixed.

I made two PRs:
- 26900 is the PR for the proposed change in the code of the class KanesMethod. It passed the Github tests without problems.
- 26901 (this one) is the PR for the test function for the change of the code in KanesMethod.

Should this have been one PR?
**Thanks!**

**Comment by moorepants** [comment]
A single PR should contain both the changes to the code and a unit test to show your changes work. You only have the latter in this PR.

**Comment by Peter230655** [comment]
> A single PR should contain both the changes to the code and a unit test to show your changes work. You only have the latter in this PR.

So I will make a new PR containing both:
- the proposed change in the code for KanesMethod
- the test for this

**Thanks!**

**Comment by moorepants** [comment]
No, just fix this one!

**Comment by Peter230655** [comment]
> No, just fix this one!

Do you mean, I add the proposed change in KanesMethod to this branch and make a new PR?

**Comment by moorepants** [comment]
You have an open PR here (26901) that is associated with your local branch named `test_linearity_of_speed...`.  It only has the unit test as a change. Now add the code changes to `kane.py`. Don't make a new branch. Don't make a new PR. Just edit the files in this branch, commit, and push.

**Comment by Peter230655** [comment]
> You have an open PR here (26901) that is associated with your local branch named `test_linearity_of_speed...`. It only has the unit test as a change. Now add the code changes to `kane.py`. Don't make a new branch. Don't make a new PR. Just edit the files in this branch, commit, and push.

Got it! (Sorry I am so dumb with this stuff! Easier for me to write 5 simulations with opty than getting this Github stuff right!)

**Comment by tjstienstra** [comment]
Might be better to use `linear_eq_to_matrix` when computing `f_nh` and `k_nh`. See also #25070

**Comment by Peter230655** [comment]
I added the test for nonlinearity to PR 26900 where I propose the change in kane.py. Consequently, I will close this PR.
